ODPi-Runtime.md

```diff
@@ -225,33 +225,194 @@ Best practices for ODPi Platforms:
+
+in which the HCFS FileSystem implementation class is named.  (For an example, see [AWS S3 FileSystem implementation](https://github.com/apache/hadoop/blob/trunk/hadoop-tools/hadoop-aws/src/main/resources/META-INF/services/org.apache.hadoop.fs.FileSystem))
+
+There is no constraint on the naming of the AbstractFileSystem implementation class, but by convention it also SHOULD be in the 
+package `org.apache.hadoop.fs.<hcfsname>`.  It SHOULD NOT be named in the `META-INF/services` file.
+
+An HCFS implementation MUST satisfy all of the API contracts outlined in the [*Hadoop FileSystem API Definition*](http://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/index.html)
+
+All behavior semantics of the HDFS implementation SHOULD be imitated insofar as the file system underlying the HCFS is compatible with 
+such behavior.  If the HCFS is capable of providing locality information, it SHOULD support the API for 
+telling applications such as Apache Tez and Apache Spark which TaskTracker is closest to the 
+data (`FileSystem:getFileBlockLocations()`).  Where HDFS-native APIs cannot be implemented, then the HCFS SHOULD return values 
+which are compatible with the rest of the Hadoop stack. As an example, while getBlockSize() is not a value which a blockless 
+filesystem can actually return, a value suitable for block-partitioning algorithms should be returned, such as 256M, 512M or 
+similar.
+
+Almost any file system can be interfaced to the Hadoop stack via the HCFS interface, and can therefore be used for basic I/O and 
+mass storage, alongside HDFS.  They are still HCFS implementations, but additional features (consistency, atomicity, and durability) 
+MUST be provided in order for the file system to serve as primary storage replacing HDFS in its role as primary storage and working
+space for MapReduce, Tez, and HBase. The difference is discussed in detail in the section titled [*Object Stores vs Filesystems*](http://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/introduction.html#Object_Stores_vs._Filesystems)
+
+As explained there, the key properties of consistency, atomicity, and durability, are missing from many filesystem-like storage 
+platforms, including Amazon S3 and OpenStack Swift.  These are often referred to as Object Stores or Blob Stores, but be aware 
+that this distinction by name is not necessarily correct.  In particular, some Blob Store-based file systems *do* support the 
+necessary features and can be substituted for HDFS.  For example, on the Microsoft Azure HDInsights platform, the WASB file 
+system (“Windows Azure Storage - Blob”) supports all features necessary to act as primary storage and working space for 
+MapReduce, Tez, and HBase, and is successfully substituted for HDFS on the HDInsights platform. Regardless of whether an HCFS
+implementation provides a drop-in replacement for HDFS it MUST support all of the ODPi Core components without degradation in
```

Comment:
Regarding the last sentence ("Regardless of whether..."), must distinguish between HCFS supported by the ODPi Platform vendor for substitution in place of HDFS, vs HCFS supported only for use via generic input/output.  Some core components, including MapReduce, Tez and HBase, will not be fully supported by the latter.  Suggest the phrasing:
* If an ODPi Platform supports an HCFS Filesystem Product as a drop-in replacement for HDFS, then it MUST support all of the ODPi Core components without degradation in functionality.  If an ODPi Platform supports an HCFS Filesystem Product only for use with Application input/output, then it must support Applications' use of FileContext APIs without degradation in functionality.